libs/execution/src/lib/hooks/hook-context.ts

```diff
@@ -0,0 +1,195 @@
+// SPDX-FileCopyrightText: 2025 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+// eslint-disable-next-line unicorn/prefer-node-protocol
+import assert from 'assert';
+
+import { type Result } from '../blocks';
+import { type ExecutionContext } from '../execution-context';
+import { type IOTypeImplementation } from '../types';
+
+import {
+  type HookOptions,
+  type HookPosition,
+  type PostBlockHook,
+  type PreBlockHook,
+  isPreBlockHook,
+} from './hook';
+
+const AllBlocks = '*';
+
+interface HookSpec<H extends PreBlockHook | PostBlockHook> {
+  blocking: boolean;
+  hook: H;
+}
+
+async function executeTheseHooks(
```

Comment:
I think it's better so split these functions into two. The typing is quite complex and will get more complex the more functionality we add. 
When we add the functionality to alter the input/output value, the logic will be too different as well I think.